processor/src/chiplets/aux_trace/mod.rs

```diff
@@ -245,3 +285,728 @@ impl AuxColumnBuilder<ChipletsVTableUpdate, ChipletsVTableRow, u32> for Chiplets
+                + alphas[1].mul_base(op_label)
+                + alphas[2].mul_base(row[DECODER_HASHER_RANGE.start])
+                + alphas[3].mul_base(row[DECODER_HASHER_RANGE.start + 1])
+                + alphas[4].mul_base(row[DECODER_HASHER_RANGE.start + 2])
+                + alphas[5].mul_base(row[DECODER_HASHER_RANGE.start + 3]);
+            multiplicand *= factor;
+        }
+        multiplicand *= factor;
+    }
+
+    // RESPAN block
+    if [b0, b1, b2, b3, b4, b5, b6] == [ZERO, ZERO, ZERO, ONE, ONE, ONE, ONE] {
+        let op_label = LINEAR_HASH_LABEL;
+
+        let first_cycle_row = addr_to_row_index(col_addr[i + 1] - ONE) % 8 == 0;
+        let transition_label = if first_cycle_row { op_label + 16 } else { op_label + 32 };
+
+        let header = alphas[0]
+            + alphas[1].mul_base(Felt::from(transition_label))
+            + alphas[2].mul_base(col_addr[i + 1] - ONE)
+            + alphas[3].mul_base(ZERO);
+
+        // TODO: access only relevant portion of trace
+        let mut row = vec![ZERO; TRACE_WIDTH];
+        main_trace.read_row_into(i - 2, &mut row);
+
+        let mut row_next = vec![ZERO; TRACE_WIDTH];
+        main_trace.read_row_into(i - 1, &mut row_next);
+
+        let state = &row[HASHER_RATE_COL_RANGE];
+        let state_nxt = &row_next[HASHER_RATE_COL_RANGE];
+
+        let factor =
+            header + build_value(&alphas[8..16], state_nxt) - build_value(&alphas[8..16], state);
+
+        multiplicand *= factor;
+    }
+
+    // END of block
+    if [b0, b1, b2, b3, b4, b5, b6] == [ZERO, ZERO, ZERO, ZERO, ONE, ONE, ONE] {
+        let op_label = RETURN_HASH_LABEL;
+        let first_cycle_row = addr_to_row_index(col_addr[i] + Felt::from(7_u64)) % 8 == 0;
+        let transition_label = if first_cycle_row { op_label + 16 } else { op_label + 32 };
+
+        let header = alphas[0]
+            + alphas[1].mul_base(Felt::from(transition_label))
+            + alphas[2].mul_base(col_addr[i] + Felt::from(7_u64))
+            + alphas[3].mul_base(ZERO);
+
+        // TODO: access only relevant portion of trace
+        let mut row = vec![ZERO; TRACE_WIDTH];
+        main_trace.read_row_into(i, &mut row);
+
+        let state = &row[DECODER_HASHER_RANGE.start..DECODER_HASHER_RANGE.start + 4];
+
+        let factor = header + build_value(&alphas[8..12], state);
+        multiplicand *= factor;
+    }
+
+    // U32AND or U32XOR
+    if [b1, b2, b3, b4, b5, b6] == [ONE, ONE, ZERO, ZERO, ONE, ZERO] {
+        let s0_col = main_trace.get_column(STACK_TRACE_OFFSET);
+        let s1_col = main_trace.get_column(STACK_TRACE_OFFSET + 1);
+
+        let [s0, s1, s2, s3] = [ONE, ZERO, b0, ZERO];
+        let op_label = get_op_label(s0, s1, s2, s3);
+
+        let factor = alphas[0]
+            + alphas[1].mul_base(op_label)
+            + alphas[2].mul_base(s1_col[i])
+            + alphas[3].mul_base(s0_col[i])
+            + alphas[4].mul_base(s0_col[i + 1]);
+        multiplicand *= factor;
+    }
```

Comment:
I would probably create helper function for individual requests and responses (in the future, we may even move them into their relative modules - but for now, let's keep them in this file). For example:
```Rust
fn build_bitwise_request<E: FieldElement<BaseField = Felt>>(
    main_trace: &MainTrace,
    is_xor: Felt,
    alphas: &[E],
    i: usize
) -> E {
    let op_label = get_op_label(ONE, ZERO, is_xor, ZERO);
    let a = main_trace.get_s0(i);
    let b = main_trace.get_s1(i);
    let z = main_trace.get_s0(i + 1);

    alphas[0]
        + alphas[1].mul_base(op_label)
        + alphas[2].mul_base(a)
        + alphas[3].mul_base(b)
        + alphas[4].mul_base(z)
}
```
And in the match arms, I would handle them explicitly:
```Rust
let multiplicand = match op_code {
    ...
    U32AND => build_bitwise_request(&main_trace, ZERO, alphas, i),
    U32XOR => build_bitwise_request(&main_trace, ONE, alphas, i),
    ...
};
```

Reply:
Done! Sometimes having the op code as a Felt can help avoiding some back and forth conversion, so I have added it as an argument to some of the helper build functions. Let me know if this should be modified.